sourcegraph/sourcegraph
# [CLOUD-222] auth: account locked out after consecutive failed attempts

This PR implements account lockout for the builtin auth provider, currently accounts will be locked out after 5 consecutive failed attempts within an hour, and the lockout period is 30 minutes.

## Test plan

Unit tests and,

1. Boot up local instance (doesn't have to be in dotcom mode)
2. Try wrong password for an existing user for 5 times
3. On the sixth time, the account lockout error is shown
4. Try again after 30 minutes (😂) or delete the Redis key `v2:account_lockout:<user ID>`, the account is unlocked

<img width="429" alt="CleanShot 2022-04-08 at 16 22 57@2x" src="https://user-images.githubusercontent.com/2946214/162395810-7ec20d83-70ea-467e-9cbd-34172a21b1cc.png">


---

Jira: CLOUD-222
